### PR TITLE
Add authentication history

### PR DESCRIPTION
**Pull request by Gargron**
Keep track of individual sign-in events, like when someone signed-in through a successful e-mail security code, or just password alone, from which IP and browser (only keep that as long as other IPs are retained). Should help detecting/troubleshooting unauthorized access.

image

### CODE REVIEW

File: app/models/status.rb [hunk 1]
```diff
@@ -23,6 +23,9 @@
 #  in_reply_to_account_id :bigint(8)
 #  poll_id                :bigint(8)
 #  deleted_at             :datetime
+#  longitude              :decimal(, )
+#  latitude               :decimal(, )
+#  place_name             :string
```
Comment (ClearlyClaire) [inline app/models/status.rb:28]: Unrelated changes.

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]

**Comment by Gargron** [comment]
Decided not to add it to admin UI in this PR since I noticed the way recent IPs are displayed in the admin UI needs a bit of a revamp...

File: app/views/settings/login_activities/_login_activity.html.haml [hunk 2]
```diff
@@ -0,0 +1,17 @@
+- method_str = content_tag(:span, login_activity.omniauth? ? t(login_activity.provider, scope: 'auth.providers') : t(login_activity.authentication_method, scope: 'login_activities.authentication_methods'), class: 'target')
+- ip_str = content_tag(:span, login_activity.ip, class: 'target')
+- browser_str = content_tag(:span, t('sessions.description', browser: t("sessions.browsers.#{login_activity.browser}", default: "#{login_activity.browser}"), platform: t("sessions.platforms.#{login_activity.platform}", default: "#{login_activity.platform}")), class: 'target')
```
Comment (rinsuki) [inline app/views/settings/login_activities/_login_activity.html.haml:3]: I want to see full user-agent in `title` attribute, like active sessions list:
image

File: app/controllers/auth/sessions_controller.rb [hunk 3]
```diff
@@ -136,4 +136,34 @@ def clear_attempt_from_session
...
+  end
+
+  def on_authentication_failure(user, security_measure, failure_reason)
+    LoginActivity.create(
+      user: user,
+      success: false,
+      authentication_method: security_measure,
+      failure_reason: failure_reason,
+      ip: request.remote_ip,
+      user_agent: request.user_agent
```
Comment (ClearlyClaire) [inline app/controllers/auth/sessions_controller.rb:166]: This only pops up after correct password input, right? If not, I feel like in many cases it could just be noise with no reasonable action to take from the user's perspective.

File: app/controllers/auth/sessions_controller.rb [hunk 4]
```diff
@@ -25,9 +25,11 @@ def new
...
   def create
     super do |resource|
-      resource.update_sign_in!(request, new_sign_in: true)
-      remember_me(resource)
-      flash.delete(:notice)
+      # We only need to call this if this hasn't already been
+      # called from one of the two-factor or sign-in token
+      # authentication methods
+
+      on_authentication_success(resource, :password) unless @on_authentication_success_called
```
Comment (ClearlyClaire) [inline app/controllers/auth/sessions_controller.rb:32]: Interestingly, this didn't call `sign_in(user)` but `on_authentication_success` now does. Could that become an issue?

File: app/controllers/auth/sessions_controller.rb (hunk 3 above)
Comment (Gargron) [reply app/controllers/auth/sessions_controller.rb:166]: Indeed. I decided to ignore incorrect password inputs since there would be too many. But if someone has your password and is trying to break 2fa, it could be useful to know.

File: app/controllers/auth/sessions_controller.rb (hunk 4 above)
Comment (ClearlyClaire) [reply app/controllers/auth/sessions_controller.rb:32]: So, looked into it a bit more: `sign_in` is already called in `Devise::SessionsController#create`, so this call is superfluous but it doesn't seem like it could cause issues.

**Review by ClearlyClaire — APPROVED** [review]

**Comment by ClearlyClaire** [comment]
Regarding logging authentication errors… I'm not sure how useful that is, as I expect an attacker would just… not enter a 2FA code once they got the password prompt right? Maybe it would make sense to log in successful password prompts that did not lead to a successful login because of 2FA?

**Comment by Izorkin** [comment]
cc @Gargron @ClearlyClaire 
I have a network with subnet mask 2001:db8:0021::/64 (example).
Every time the IPv6 address changes dynamically, and when you log into your account again, protection is triggered, because. the IPv6 address has changed.
Is it possible for an IPv6 address to be ignored by the /64 mask?
